modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/UpdateDataStreamGlobalRetentionService.java

```diff
@@ -0,0 +1,165 @@
+import org.elasticsearch.datastreams.lifecycle.action.PutDataStreamGlobalRetentionAction;
+import org.elasticsearch.datastreams.lifecycle.action.UpdateDataStreamGlobalRetentionResponse;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+public class UpdateDataStreamGlobalRetentionService {
+
+    private final MasterServiceTaskQueue<UpsertGlobalDataStreamMetadataTask> taskQueue;
+
+    public UpdateDataStreamGlobalRetentionService(ClusterService clusterService) {
+        ClusterStateTaskExecutor<UpsertGlobalDataStreamMetadataTask> executor = new SimpleBatchedAckListenerTaskExecutor<>() {
+
+            @Override
+            public Tuple<ClusterState, ClusterStateAckListener> executeTask(
+                UpsertGlobalDataStreamMetadataTask task,
+                ClusterState clusterState
+            ) {
+                return new Tuple<>(updateGlobalRetention(clusterState, task.globalRetention()), task);
+            }
+        };
+        this.taskQueue = clusterService.createTaskQueue("data-stream-global-retention", Priority.HIGH, executor);
+
+    }
+
+    public void updateGlobalRetention(
+        PutDataStreamGlobalRetentionAction.Request request,
+        List<UpdateDataStreamGlobalRetentionResponse.AffectedDataStream> affectedDataStreams,
+        final ActionListener<UpdateDataStreamGlobalRetentionResponse> listener
+    ) {
+        taskQueue.submitTask(
+            "update-data-stream-global-retention",
+            new UpsertGlobalDataStreamMetadataTask(
+                request.getGlobalRetention(),
+                affectedDataStreams,
+                listener,
+                request.masterNodeTimeout()
+            ),
+            request.masterNodeTimeout()
+        );
+    }
+
+    public void removeGlobalRetention(
+        DeleteDataStreamGlobalRetentionAction.Request request,
+        List<UpdateDataStreamGlobalRetentionResponse.AffectedDataStream> affectedDataStreams,
+        final ActionListener<UpdateDataStreamGlobalRetentionResponse> listener
+    ) {
+        taskQueue.submitTask(
+            "remove-data-stream-global-retention",
+            new UpsertGlobalDataStreamMetadataTask(null, affectedDataStreams, listener, request.masterNodeTimeout()),
+            request.masterNodeTimeout()
+        );
+    }
+
+    public List<UpdateDataStreamGlobalRetentionResponse.AffectedDataStream> determineAffectedDataStreams(
+        @Nullable DataStreamGlobalRetention newGlobalRetention,
+        ClusterState clusterState
+    ) {
+        var previousGlobalRetention = DataStreamGlobalRetention.getFromClusterState(clusterState);
+        if (Objects.equals(newGlobalRetention, previousGlobalRetention)) {
+            return List.of();
+        }
+        List<UpdateDataStreamGlobalRetentionResponse.AffectedDataStream> affectedDataStreams = new ArrayList<>();
+        for (DataStream dataStream : clusterState.metadata().dataStreams().values()) {
```

Comment:
This is going to run on the transport thread on the master node, so it needs to be fast. I think this is plenty fast, but I'm just pointing it out in case someone thinks that we might have so many data streams in a large cluster that we'd want to put this for loop in another thread.

Reply:
Ignore this comment -- it is not on the transport thread after all! It's running in a management thread.